### DISCUSSION TITLE
What does it mean when the confusion matrix points to "background"?

### DISCUSSION THREAD

**Discussion by JohannesBauer97** [ultralytics/yolov5#9907] [Q&A]
I trained my model using a custom dataset for laundry symbols and it stopped after 175 epochs due overfitting.
My confusion matrix also does not look really healthy, it's pointing to "background".

Does this mean my dataset is too small and the model still cannot understand the difference between object and background?

````
!python CareSymbols/yolov5/train.py --img 1280 --batch -1 --epochs 300 --data care.yaml --weights yolov5s6.pt --cache ram --patience 15
````

![confusion_matrix](https://user-images.githubusercontent.com/15695124/197554107-7411789e-1adf-4776-bda5-f68bf1621925.png)

![labels](https://user-images.githubusercontent.com/15695124/197554389-ccdd5ce7-2cfe-4ce3-a9a5-8a8fd61be9bc.jpg)

![results](https://user-images.githubusercontent.com/15695124/197554418-44f7db59-e3c1-406d-bdd5-60912fdb8694.png)
![train_batch0](https://user-images.githubusercontent.com/15695124/197554528-1b29213a-994b-4fd5-8e32-eb44e54d30f7.jpg)

**Comment by glenn-jocher** [comment]
@JohannesBauer97 your model is predicting background instead of detecting objects.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by TheArmbreaker** [comment]
Hello @JohannesBauer97,

First I think it is important to mention that due to [Issue 2071](https://github.com/ultralytics/yolov5/issues/2071) the vectors for False Positive and False Negative are switched. Thus you are dealing with Type 2 errors (False Negative).

Therefore my suggestion would be to reduce FN and thereby improving Recall.
This might be done by teaching ground truth.

From your provided images I derived that you might provide the icons only once!?

Maybe you can
- add more pure images of the pure labeled object or some images with easier background (referring to Image variety from @glenn-jocher 's answer)
- or change the IoU Threshold for Training.

The IoU and the confidence are deciding about classification. Thus an increased IoU threshold while training (in your training hyp yaml) might force the model to look more precise.

In my currently ongoing learning project I had a lot of False Positive which disappeared with additional background images. Thus my model was detecting objects instead of background. However, I used images with easy and difficult background to reduce False Negativ.

I am still trying to figure stuff out though. Let me know, if this was of any help.

**Reply by JohannesBauer97** [reply]
I'll experiment with the IoU parameter, big thanks!
The best case would be that the model doesn't care about background and if there are symbols once or multiple times on one image.

I did another test using only white backgrounds, with the intention "give that model something easy todo".
So I generated 6k images with a white background and the model performed similar bad - no improvement at all.
Looks like I'm doing something really wrong when trying to build a model to detect 2D symbols on images. :(


**Example:**
![image5888](https://user-images.githubusercontent.com/15695124/198869906-c87fcb53-4c45-490e-a513-05845e114059.jpg)

**Reply by TheArmbreaker** [reply]
Meanwhile I figured out that the val.py is using a very-low conf-threshold (0.001) to feed a valid confusion matrix.
This automatically runs at the end of the training.

Try *!python val.py --conf 0.25* on your best.pt, this might "sort out" misclassification.
At least further of my false positives got "sorted out" for me.
_Edit: This does not work on False Negativ, because Recall is increasing with increased conf-threshold._ 

Maybe you also want to freeze the backbone in the first training to utilize pre-trained weights. I am working with guidance from this [Practical Guide for Object Detection on YOLOv5](https://towardsdatascience.com/the-practical-guide-for-object-detection-with-yolov5-algorithm-74c04aac4843). Freezing the background in a first run really made a difference for me.

Additionally the icons may be very small on very large images !?
Thus the re-scalling with size 1280 might shrink them to oblivion.

Good Luck.

**Reply by JohannesBauer97** [reply]
In my training I use -img 1280 to keep the big image size and not letting yolo shrinking the image - otherwise symbols would be hard to detect already for a human eye.
